platform/android/src/jni.cpp

```diff
@@ -373,6 +435,77 @@ std::pair<mbgl::AnnotationSegment, mbgl::ShapeAnnotation::Properties> annotation
     return std::make_pair(segment, shapeProperties);
 }
 
+/*
+ * Adapted from https://github.com/dropbox/djinni/blob/master/support-lib/jni/Marshal.hpp
+ */
+
+void jniExceptionCheck(JNIEnv * env) {
+    if (!env) {
+        mbgl::Log::Error(mbgl::Event::JNI, "Empty environment.");
+        abort();
+    }
+
+    const jthrowable e(env->ExceptionOccurred());
+    if (e) {
+        env->ExceptionClear();
```

Comment:
This doesn't look like the desired semantics given how this function is used. This definition will clear any pending Java exception and allow C++ code that calls it to continue as if nothing has gone wrong.

In most cases, the desired behavior if a Java exception is detected is that C++ code unwinds, without doing further work, back to the JVM boundary. And then after returning to JVM control, Java exception handling takes over, processing the still-pending Java exception.

This can be accomplished by throwing an exception of a specific type, say `PendingJavaException`, here, and catching and discarding it in all native binding methods just before returning to JVM control. This is the approach that the JNI library I'm writing will take.